Zend/zend_operators.c

```diff
@@ -78,58 +80,72 @@ static const unsigned char tolower_map[256] = {
 		zend_binary_strncasecmp
  */
 
-ZEND_API int ZEND_FASTCALL zend_atoi(const char *str, size_t str_len) /* {{{ */
+static void zend_atol_warning(const char *str, size_t str_len, size_t num_len, char suffix) /* {{{ */
 {
-	int retval;
+	/* Duplicate just in case str[str_len] != 0 */
+	zend_string *strcopy = zend_string_init(str, str_len, 0);
+	zend_string *numonly = zend_string_init(str, num_len, 0);
+	const char suffixstr[2] = { suffix, '\0' };
 
-	if (!str_len) {
-		str_len = strlen(str);
-	}
-	retval = ZEND_STRTOL(str, NULL, 0);
-	if (str_len>0) {
-		switch (str[str_len-1]) {
-			case 'g':
-			case 'G':
-				retval *= 1024;
-				/* break intentionally missing */
-			case 'm':
-			case 'M':
-				retval *= 1024;
-				/* break intentionally missing */
-			case 'k':
-			case 'K':
-				retval *= 1024;
-				break;
-		}
-	}
-	return retval;
+	zend_error(E_WARNING, "Invalid numeric string '%s', interpreting as '%s%s'", ZSTR_VAL(strcopy), ZSTR_VAL(numonly), suffixstr);
```

Comment:
Maybe use `%.*s` here with explicitly passed length to avoid those string copies? Can also use `%c` instead of the suffixstr.

Reply:
Ah, right. `%.*s` was the format I was trying to recall. Thanks.

On the suffixstr one, I don't want an empty space introduced when it's `\0`.